simularium/nbsv
# Feature/checkboxes

## Agent Checkbox Tree

Closes #13 
Design review with Travis complete.
_Review time: long, lots of ui, functions, typing_

Re-opening a very changed PR that uses selectors and React Context.

The UI displays a tree of agents and subagents, each with their own row. In each row are two checkboxes (but one looks like a star. Clicking the conventional checkbox handles hiding agents, which default to being shown, and clicking the star handles highlighting gents.

Besides the obvious behavior of hiding/highlighting. Parent agents and the "all agents" level should be in `Indeterminate` state if any of their children are in different states. Clicking an indeterminate button should make everything below that level in the tree `Inactive` i.e. not highlighted and not hidden.

Demo:

https://github.com/simularium/nbsv/assets/24981838/42b33318-0b58-44ad-87a7-5f7b9ae6f521

Approach
------------------

Designating hiding and highlighting, or sometimes for choosing between a set of hidden or highlighted agents makes use of:
```
enum DisplayAction {
  Hide = 'hidden',
  Highlight = 'highlight',
}
```
To the sim-viewer, an agent or displayState (aka subAgent) are either hidden/highlighted, or they are not. But in the UI, a level can be indeterminate if any of its children are in different states, so:
```
enum HiddenOrHighlightedState {
  Active = 'Active',
  Inactive = 'Inactive',
  Indeterminate = 'Indeterminate',
}
```
is useful in both functions and for mapping icons and text into generalized pieces of the UI.

We've avoided using Redux this far, but a minimal bit of state management via `useContext` seemed sensible to prevent repetition and prop drilling while pulling selectors out into their own file.

This app tracks hidden and highlighted state in `currentVisibilityStates` which are each a `VisibilitySelectionMap` provided to the component tree by `VisibilityContext`.

`selectionStateInfo` for the viewer is derived from that, and the context also provides the functions that update state when checkboxes are clicked.

The `subAgentDisplayMaps` are agent specific and that state is held in `CheckBoxContents` but the function to update those maps is also down from the context provider, so that the context provider is the only place that the selectors are imported and used.

## Key fIles:

AgentVisibilityContext.tsx
-----------------
A React context provider that holds and provides the data and functions relevant to agent visibility. Has a type, a set of default values, and then the `VisibilityProvider` holds the actual functions and state variables.

Widget.tsx
-----------------
To provide the context to all layers of the tree including `Viewer` I had to wrap everything in the VisibilityProvider which looks a little awkward given the way we create the element in the widget:
```
    const wrappedComponent = React.createElement(
      VisibilityProvider,
      null,
      React.createElement(Viewer, {
        controller: this.controller,
      } as ViewerProps)
    );

    ReactDOM.render(wrappedComponent, this.el);
```

Viewer.tsx
-------------------
Receives the `uiDisplayData` to send to context and sends the `SelectionStateInfo` from context to sim-viewer.

SidePanel.tsx
-------------------
Renders the `CheckBoxContents` but also handles hiding/revealing all agents at once.

CheckBoxContents.tsx
-------------------
This is the first `agent` specific component.  We keep the `topLevelDisplayStatus` and `subAgentDisplayMaps` here and watch for visibility changes with useEffect. Trying to centralize this state in the context would be frustrating because it's all initially undefined until agents are loaded with the trajectory.

CheckBoxRow.tsx
-------------------
Mostly display and passing the appropriate arguments into the display action handlers.


Steps to Verify:
----------------
- run the viewer
- confirm that the agents and the subagents each have a row
- when you click a star it should highlight/remove highlight for that agent or set of agents
- when you click a check box it should hide/restore the visibility of that agent or group of agents
- if an agent is indeterminate (its children don't share their hidden/highlighted status), or when using the all agent types checkbox: clicking the top level check box should remove highlights, and reveal/unhide all children agents
